File: src/main/java/introkotlin/samplejava/step6/Money.java
package introkotlin.samplejava.step6;

import java.util.Objects;

/**
 * Aucun changement par rapport au step 5.
 */
public final class Money {

    private final double value;

    private final String currency;

    public Money(double value, String currency) {
        this.value = value;
        this.currency = currency;
    }

    public double getValue() {
        return value;
    }

    public String getCurrency() {
        return currency;
    }

    public Money plus(Money other) {
        if (!currency.equals(other.currency)) {
            throw new IllegalArgumentException("Different currencies: " + currency + " != " + other.currency);
        }
        return new Money(value + other.value, currency);
    }

    @Override
    public String toString() {
        return "Money(value=" + value + ", currency=" + currency + ")";
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        final var other = (Money) obj;
        return other.value == value && Objects.equals(currency, other.currency);
    }

    @Override
    public int hashCode() {
        return Objects.hash(value, currency);
    }

}

File: src/main/kotlin/introkotlin/samplekotlin/step6/Money.kt
/**
 * Démonstration des extension methods.
 */
package introkotlin.samplekotlin.step6

import java.util.concurrent.ThreadLocalRandom

data class Money(
  val value: Double,
  val currency: String
) {

  operator fun plus(other: Money): Money {
    if (currency != other.currency) {
      throw IllegalArgumentException("Different currencies: $currency != ${other.currency}")
    }
    // copy : méthode introduite par le caractère "data" de la classe
    // On peut également nommer les arguments des méthodes, pour mieux
    // s'y retrouver par la suite
    return copy(value = value + other.value)
  }

}

/**
 * On peut enrichir des objets avec des extension methods.
 * C'est un moyen simple d'enrichir un objet avec des
 * méthodes utilitaires.
 */
fun Money.toBitcoin(): Money {
  val exchangeRate = ThreadLocalRandom.current().nextDouble(0.0, 0.001)
  return Money(value = value * exchangeRate, currency = "BITCOIN")
}

fun main(vararg args: String) {
  repeat(10) {
    val money = Money(10.0, "EURO")
    System.out.println("Money = ${money.toBitcoin()}")
  }
}
